cc/main.cc

```diff
@@ -464,46 +467,244 @@ class SelfPlayer {
   uint64_t flags_timestamp_ = 0;
 };
 
+class Evaluator {
+  // A barrier that blocks threads until the number of waiting threads reaches
+  // the 'count' threshold. Allows decrementing the threshold to handle the tail
+  // of a work queue where some threads exit early.
+  class Barrier {
+   public:
+    explicit Barrier(size_t count)
+        : count_(count), num_waiting_(0), generation_(0) {}
+
+    void Wait() {
+      absl::MutexLock lock(&mutex_);
+      if (++num_waiting_ == count_) {
+        IncrementGeneration();
+      } else {
+        auto generation = generation_;
+        while (generation != generation_) {
+          cond_var_.Wait(&mutex_);
+        }
+      }
+    }
+
+    void DecrementCount() {
+      absl::MutexLock lock(&mutex_);
+      if (num_waiting_ == --count_) {
+        IncrementGeneration();
+      }
+    }
+
+   private:
+    void IncrementGeneration() EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+      ++generation_;
+      num_waiting_ = 0;
+      cond_var_.SignalAll();
+    }
+
+    absl::Mutex mutex_;
+    absl::CondVar cond_var_;
+    size_t count_ GUARDED_BY(&mutex_);
+    size_t num_waiting_ GUARDED_BY(&mutex_);
+    size_t generation_ GUARDED_BY(&mutex_);
+  };
+
+  // References a pointer to an actual DualNet. Allows updating the pointer
+  // after the MctsPlayer has been constructed.
+  class WrappedDualNet : public DualNet {
+   public:
+    explicit WrappedDualNet(const std::unique_ptr<DualNet>& dual_net)
```

Comment:
const std::unique_ptr<DualNet>* please. See the internal C++ Tip of the Week 116 for more details.

To be honest, the necessity of this class strongly suggests to me that the current DualNet & factory design is no longer working for your requirements. As discussed offline, let's go ahead with this for now because of the tight schedule and we'll revisit soon when we implement eval on Cloud TPU.

Reply:
We might indeed need to reintroduce the Client ;-)